TechnicalDocuments/ControlTheWorld.md

```diff
@@ -0,0 +1,139 @@
+
+How to Control the World
+========================
+
+## Introduction
+
+This proposal is heavily based on episodes [16](https://www.pointfree.co/episodes/ep16-dependency-injection-made-easy) and [18](https://www.pointfree.co/episodes/ep18-dependency-injection-made-comfortable) of [Pointfree](https://www.pointfree.co) and Stephen Celis's talk [How to Control the World](https://vimeo.com/291588126) at NSSpain 2018. Slides of that talk are [also available](https://speakerdeck.com/stephencelis/how-to-control-the-world).
+
+## Motivation
+
+To overcome some of the disadvantages of dependency injection:
+
+* Adding a new dependency down the chain can mean passing that dependency all the way down through many layers. Many of those layers only need to get passed the dependency in order to pass it down to the next layer. This leads to a lot of source files being changed just to add a single dependency.
+
+* Many of our dependencies, such as the Analytics or the AB Testing Service are, in reality, singletons which never change at runtime and are just passed all over the place.
+
+* In order to gain the testing benefit of dependency injection (e.g: being able to substitute a dependency with a stub implementation for testing), you have to use Protocols, which add unnecessary boilerplate.
+
+## Benefits
+
+* Avoid having to pass singleton dependencies down through multiple layers of injection.
+
+* Ability to easily create different app "environments" to simulate running on different dates, timezones and locales, for example.
+
+* Ability to have "mock" environments for testing. For example, testing that analytics events are correctly being sent is typically quite difficult. With a mock analytics service in a mock app environment this becomes trivial, and without the need to inject an analytics service protocol to every part of the app.
+
+## Implementation
```

Comment:
I am not sure if the `Analytics` example should be included here, or should it at least be clarified as an example but not as part of the proposal.

Reply:
I've added a not about this.